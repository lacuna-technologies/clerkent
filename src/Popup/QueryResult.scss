#result {
  display: flex;
  flex-direction: column;
  margin-top: 0.25rem;

  > .jurisdiction {
    text-transform: uppercase;
    margin-bottom: 0.3rem;
  }

  > .links {
    margin-top: 0.3rem;

    > span + a {
      margin-left: 0.5rem;
    }
  }

  > p {
    margin: 0;
  }
}