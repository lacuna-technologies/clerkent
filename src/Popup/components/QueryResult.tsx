import { useCallback, useState } from 'preact/hooks'
import CaseResult from './CaseResult'
import LegislationResult from './LegislationResult'
import ShowMore from './ShowMore'
import AnimatedLoading from '../../components/AnimatedLoading'
import type { FunctionComponent } from 'preact'

interface Props {
  searchResult: (Law.Case | Law.Legislation)[],
  downloadPDF: downloadPDFType,
  isSearching: boolean,
}

const maxResults = 3

// eslint-disable-next-line sonarjs/cognitive-complexity
const QueryResult: FunctionComponent<Props> = ({ searchResult, downloadPDF, isSearching }) => {
  const [morePressed, setMorePressed] = useState(false)
  const onShowMore = useCallback(() => setMorePressed(true), [])

  if (isSearching){
    return <AnimatedLoading />
  }

  if(searchResult.length === 0){
    return <div className="flex-grow">No cases found</div>
  }

  const showMore = morePressed || searchResult.length <= maxResults

  const resultType: Law.Type = searchResult[0]?.type
  if(resultType === `case-citation` || resultType === `case-name`){
    return (
      <div className="flex flex-col mt-2 items-start content-start gap-4">
        {
          (searchResult as Law.Case[])
            .slice(0, showMore ? undefined : maxResults)
            .map((result) => (
              <CaseResult
                case={result}
                downloadPDF={downloadPDF}
                key={`${result.name}-${result.citation}`}
              />
            ))
        }
        { showMore ? null : <ShowMore onClick={onShowMore} /> }
      </div>
    )
  } else if (resultType === `legislation`){
    return (
      <div className="flex flex-col mt-2 items-start content-start gap-4">
        {
          (searchResult as Law.Legislation[])
          .slice(0, showMore ? undefined: maxResults)
          .map((legislation) => {
            const { provisionType, provisionNumber, statute } = legislation
            return (
              <LegislationResult
                legislation={legislation}
                downloadPDF={downloadPDF}
                key={`${provisionType}-${provisionNumber}-${statute}`}
              />
            )
          })
        }
        { showMore ? null : <ShowMore onClick={onShowMore} /> }
      </div>
    )
  }
  return null
}

export default QueryResult