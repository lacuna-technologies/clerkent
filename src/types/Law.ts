declare namespace Law {
  interface Database {
    icon: string,
    name: string, 
    url: string,
    id: string,
  }

  interface Link {
    doctype: `Summary` | `Judgment` | `Opinion` | `Legislation`,
    filetype: `PDF` | `HTML`,
    url: string,
  }
  interface Case {
    name: string,
    citation: string,
    links: Link[]
    jurisdiction?: JursidictionCode
    database: Database
    type?: Type
  }
  interface Legislation {
    provisionType: string,
    provisionNumber: string,
    statute: string,
    links: Link[],
    jurisdiction?: JursidictionCode,
    content?: string,
    database: Database
    type?: Type
  }

  // ISO 3166-1 alpha-2
  type JursidictionCode = `SG` | `UK` | `EU` | `HK` | `CA` | `AU` | `NZ` | `MY` | `IN`

  type SearchMode = `case` | `legislation`
  type Type = `case-citation` | `case-name` | `legislation`
}

export default Law