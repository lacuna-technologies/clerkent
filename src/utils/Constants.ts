import type Law from '../types/Law'

const JURISDICTIONS = {
  AU: {
    emoji: `ðŸ‡¦ðŸ‡º`,
    id: `AU` as Law.JursidictionCode,
    name: `Australia`,
  },
  CA: {
    emoji: `ðŸ‡¨ðŸ‡¦`,
    id: `CA` as Law.JursidictionCode,
    name: `Canada`,
  },
  EU: {
    emoji: `ðŸ‡ªðŸ‡º`,
    id: `EU` as Law.JursidictionCode,
    name: `European Union`,
  },
  HK: {
    emoji: `ðŸ‡­ðŸ‡°`,
    id: `HK` as Law.JursidictionCode,
    name: `Hong Kong`,
  },
  MY: {
    emoji: `ðŸ‡²ðŸ‡¾`,
    id: `MY` as Law.JursidictionCode,
    name: `Malaysia`,
  },
  NZ: {
    emoji: `ðŸ‡³ðŸ‡¿`,
    id: `NZ` as Law.JursidictionCode,
    name: `New Zealand`,
  },
  SG: {
    emoji: `ðŸ‡¸ðŸ‡¬`,
    id: `SG` as Law.JursidictionCode,
    name: `Singapore`,
  },
  UK: {
    emoji: `ðŸ‡¬ðŸ‡§`,
    id: `UK` as Law.JursidictionCode,
    name: `United Kingdom`,
  },
}


const SG_DATABASES = {
  lawnet: {
    icon: ``,
    name: `Lawnet`,
    url: `https://www.lawnet.sg/lawnet/web/lawnet/home`,
  },
  lexis: {
    icon: ``,
    name: `Lexread`,
    url: `https://www.lexread.lexisnexis.com/`,
  },
  sc: {
    icon: ``,
    name: `Supreme Court`,
    url: `https://www.supremecourt.gov.sg/news/supreme-court-judgments`,
  },
  slw: {
    icon: ``,
    name: `Singapore Law Watch`,
    url: `https://www.singaporelawwatch.sg/Judgments`,
  },
  sso: {
    icon: ``,
    name: `Singapore Statutes Online`,
    url: `https://sso.agc.gov.sg/`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw Asia`,
    url: `https://www.westlawasia.com/singapore/`,
  },
}

const UK_DATABASES = {
  bailii: {
    icon: ``,
    name: `BAILII`,
    url: `https://bailii.org/`,
  },
  legislation: {
    icon: ``,
    name: `legislation.gov.uk`,
    url: `https://www.legislation.gov.uk/`,
  },
  lexis: {
    icon: ``,
    name: `LexisNexis`,
    url: `http://www.lexisnexis.com/uk/legal`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw`,
    url: `https://uk.westlaw.com/Browse/Home/WestlawUK`,
  },
}

const EU_DATABASES = {
  curia: {
    icon: ``,
    name: `CURIA`,
    url: `https://curia.europa.eu/juris/recherche.jsf?language=en`,
  },
  epo: {
    icon: ``,
    name: `EPO`,
    url: `https://www.epo.org/law-practice/case-law-appeals`,
  },
}

const HK_DATABASES = {
  hkliihk: {
    icon: ``,
    name: `HKLII.hk`,
    url: `https://www.hklii.hk/`,
  },
  hkliiorg: {
    icon: ``,
    name: `HKLII.org`,
    url: `https://www.hklii.org/eng/`,
  },
}

const CA_DATABASES = {
  canlii: {
    icon: ``,
    name: `CanLII`,
    url: `https://www.canlii.org/en`,
  },
}

const AU_DATABASES = {
  austlii: {
    icon: ``,
    name: `AustLII`,
    url: `https://www8.austlii.edu.au`,
  },
}

const MISC_DATABASES = {
  commonlii: {
    icon: ``,
    name: `CommonLII`,
    url: `http://www.commonlii.org`,
  },
}

const dedupeObjects = (inputObject: Record<string, Record<string, string>>, jurisdictionCode: Law.JursidictionCode) =>
  Object.entries(inputObject)
  .reduce((accumulator, [id, object]: [string, Record<string, string>]) => ({
    ...accumulator,
    [`${jurisdictionCode}_${id}`]: object,
  }), {})

const DATABASES: Record<string, Law.Database> = {
  ...dedupeObjects(SG_DATABASES, `SG`),
  ...dedupeObjects(UK_DATABASES, `UK`),
  ...dedupeObjects(EU_DATABASES, `EU`),
  ...dedupeObjects(HK_DATABASES, `HK`),
  ...dedupeObjects(CA_DATABASES, `CA`),
  ...dedupeObjects(AU_DATABASES, `AU`),
  ...MISC_DATABASES,
}

const SG_COURTS = {
  sgca: {
    id: `SGCA`,
    name: `Court of Appeal`,
  },
  sghc: {
    id: `SGHC`,
    name: `High Court`,
  },
}

const EU_COURTS = {
  cjeu: {
    id: `CJEU`,
    name: `Court of Justice of the European Union`,
  },
  epo: {
    id: `EPO`,
    name: `European Patent Office`,
  },
}

const COURTS: Record<string, Record<string, string>> = {
  ...dedupeObjects(SG_COURTS, `SG`),
  ...dedupeObjects(EU_COURTS, `EU`),
}

const Constants = {
  COURTS,
  DATABASES,
  JURISDICTIONS,
}

export default Constants