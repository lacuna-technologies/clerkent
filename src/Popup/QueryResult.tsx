import React, { useCallback } from 'react'
import { browser } from 'webextension-polyfill-ts'
import { Constants } from '../utils'
import './QueryResult.scss'

const QueryResult = ({ parseResult, searchResult, downloadPDF, notFound }) => {

  const openTab = useCallback((link) => () => {
    browser.tabs.create({ active: true, url: link })
  }, [])

  if(!Array.isArray(parseResult) || parseResult.length === 0 || notFound){
    return <span>No results found</span>
  }

  
  if (searchResult.length === 0){
    return <span>Loading...</span>
  }

  const resultType = parseResult[0]?.type || searchResult[0]?.type
  if(resultType === `case-citation` || resultType === `case-name`){
    return (
      <div id="results">
        {
          searchResult.map(({ citation, name, link, pdf, jurisdiction, database }) => (
            <div className="result" key={`${name}-${link}`}>
              <p className="details">
                <span className="jurisdiction">{Constants.JURISDICTIONS[jurisdiction].emoji}</span>
                <span className="database">{database.name}</span>
              </p>
              <button className="case-name link" onClick={openTab(link)}>{name}</button>
              {
                pdf ? (
                  <p className="links">
                    <button className="pdf button" onClick={downloadPDF({ citation, name, pdf })}>PDF</button>
                  </p>
                ) : null
              }
            </div>
          ))
        }
      </div>
    )
  } else if (resultType === `legislation`){
    return (
      <div id="results">
        {
          searchResult.map(({
            jurisdiction,
            provisionNumber,
            provisionType,
            statute,
            link,
          }) => (
            <div className="result" key={`${provisionType}-${provisionNumber}-${statute}-${link}`}>
              <p className="details">
                { jurisdiction &&
                  <span className="jurisdiction">
                    {Constants.JURISDICTIONS[jurisdiction].emoji}
                  </span>
                }
              </p>
              <button className="legislation-name link" onClick={openTab(link)}>
                {provisionNumber
                  ? `${provisionType} ${provisionNumber}, `
                  : null
                }
                {statute}
              </button>
            </div>
          ))
        }
      </div>
    )
  }
  return null
}

export default QueryResult