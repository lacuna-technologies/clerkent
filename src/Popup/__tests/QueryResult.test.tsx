import React from 'react'
import renderer from 'react-test-renderer'
import Law from '../../types/Law'
import Constants from '../../utils/Constants'
import QueryResult from '../QueryResult'

const mockDownloadPDF = () => () => {}
const mockCases: Law.Case[] = [
  {
    citation: `[1884] EWHC 2 (QB)`,
    database: Constants.DATABASES.UK_bailii,
    jurisdiction: Constants.JURISDICTIONS.UK.id,
    links: [
      {
        doctype: `Judgment`,
        filetype: `HTML`,
        url: `http://www.bailii.org/ew/cases/EWHC/QB/1884/2.html`,
      },
      {
        doctype: `Judgment`,
        filetype: `PDF`,
        url: `http://www.bailii.org/ew/cases/EWHC/QB/1884/2.pdf`,
      },
    ],
    name: `R v Dudley and Stephens`,
    type: `case-name`,
  },
]

describe(`QueryResult`, () => {
  it(`renders no cases found`, () => {
    const tree = renderer.create(
      <QueryResult
        searchResult={[]}
        downloadPDF={mockDownloadPDF}
        isSearching={false}
        mode='case'
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders no legislation found`, () => {
    const tree = renderer.create(
      <QueryResult
        searchResult={[]}
        downloadPDF={mockDownloadPDF}
        isSearching={false}
        mode='legislation'
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders loading`, () => {
    const treeCase = renderer.create(
      <QueryResult
        searchResult={[]}
        downloadPDF={mockDownloadPDF}
        isSearching={true}
        mode='case'
      />,
    ).toJSON()
    expect(treeCase).toMatchSnapshot()

    const treeLegislation = renderer.create(
      <QueryResult
        searchResult={[]}
        downloadPDF={mockDownloadPDF}
        isSearching={true}
        mode='legislation'
      />,
    ).toJSON()
    expect(treeLegislation).toMatchSnapshot()
  })

  it(`renders list of UK cases`, () => {
    const tree = renderer.create(
      <QueryResult
        searchResult={mockCases}
        downloadPDF={mockDownloadPDF}
        isSearching={false}
        mode='case'
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})