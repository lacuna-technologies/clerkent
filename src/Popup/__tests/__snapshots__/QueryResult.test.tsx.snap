// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryResult renders list of UK cases 1`] = `
<div
  id="results"
>
  <div
    className="result"
  >
    <div
      className="details"
    >
      <div
        className="left"
      >
        <span
          className="jurisdiction"
        >
          ðŸ‡¬ðŸ‡§
        </span>
        <span
          className="database"
        >
          BAILII
        </span>
      </div>
      <div
        className="right"
      >
        <p
          className="links"
        >
          <button
            className="pdf button"
            onClick={[Function]}
          >
            PDF
          </button>
        </p>
      </div>
    </div>
    <a
      className="case-name link"
      href="http://www.bailii.org/ew/cases/EWHC/QB/1884/2.html"
      rel="noreferrer"
      target="_blank"
    >
      R v Dudley and Stephens
       
      [1884] EWHC 2 (QB)
    </a>
  </div>
</div>
`;

exports[`QueryResult renders loading 1`] = `
<div
  className="animated-loading-container"
>
  <div>
    Loading...
  </div>
  <div
    className="animated-loading"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`QueryResult renders loading 2`] = `
<div
  className="animated-loading-container"
>
  <div>
    Loading...
  </div>
  <div
    className="animated-loading"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`QueryResult renders no cases found 1`] = `
<span>
  No 
  cases
   found
</span>
`;

exports[`QueryResult renders no legislation found 1`] = `
<span>
  No 
  legislation
   found
</span>
`;
